htlcswitch/alias_store.go

```diff
@@ -0,0 +1,190 @@
+package htlcswitch
+
+import (
+	"encoding/binary"
+
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// aliasAllocBucket is a root-level bucket that stores the last alias
+	// that was allocated. It is used to allocate a new alias when
+	// requested.
+	aliasAllocBucket = []byte("alias-alloc-bucket")
+
+	// lastAliasKey is a key in the aliasAllocBucket whose value is the
+	// last allocated alias ShortChannelID. This will be updated upon calls
+	// to requestAlias.
+	lastAliasKey = []byte("last-alias-key")
+
+	// invoiceAliasBucket is a root-level bucket that stores the alias
+	// SCIDs that our peers send us in the funding_locked TLV. The keys are
+	// the ChannelID generated from the FundingOutpoint and the values are
+	// the remote peer's alias SCID.
+	invoiceAliasBucket = []byte("invoice-alias-bucket")
+
+	// byteOrder denotes the byte order of database (de)-serialization
+	// operations.
+	byteOrder = binary.BigEndian
+
+	// startingAlias is the first alias ShortChannelID that will get
+	// assigned by requestAlias. The starting BlockHeight is chosen so that
+	// legitimate SCIDs in integration tests aren't mistaken for an alias.
+	startingAlias = lnwire.ShortChannelID{
+		BlockHeight: 10000,
+		TxIndex:     0,
+		TxPosition:  0,
+	}
+)
+
+// aliasStore is a struct that has an underlying database and uses it to
+// allocate alias ShortChannelIDs. It only stores the last allocated alias. It
+// is also capable of storing the remote peer's alias SCIDs for
+// option_scid_alias channels.
+type aliasStore struct {
+	backend kvdb.Backend
+}
+
+func newAliasStore(db kvdb.Backend) *aliasStore {
+	return &aliasStore{backend: db}
+}
+
+// putPeerAlias stores the peer's alias SCID once we learn of it in the
+// funding_locked message.
+func (s *aliasStore) putPeerAlias(chanID lnwire.ChannelID,
```

Comment:
So is this a _global_ store, or somethign that's meant to be scoped per peer? If the latter, I would expect another mapping level here (pubkey -> actual set of aliases).

Reply:
It's a global store since if I give out `{10000:0:0}` to Bob to route over the me-Bob channel and `{10000:0:0}` to Carol to route over the me-Carol channel, when I receive `{10000:0:0}` from an incoming node in the onion, I won't know who to forward to